### ISSUE TITLE
Changing game region on the fly doesn't properly adjust timings

### ISSUE THREAD

**Issue by divanikus** [MiSTer-devel/Genesis_MiSTer#187]
Hi Folks!

By default Genesis core is set to detect game's region by extension. For most ROMs it ends up in EU (PAL) region for me. I thought that's not a big deal, so I was changing region manually and did a reset. 

I can confirm that game resolution changes to NTSC, so did the refresh rate (59 Hz). For games like Sonic 3 or Contra Hardcorps it is enough to go through "NTSC only game" screen and everything kind of works.

But for games like Sonic & Knuckles, Sonic 3 & Knuckles (combined ROM), may be others - it causes a ton of glitches. Messy sprites, flickering horizontal lines, CRAM dots at the top of the screen etc.

If I use header detection, the game starts in it's real region and works flawlessly.

I believe it's a bug for an end user. Because changing regions isn't locked after loading, it does changes resolution etc, but the game's engine is completely messed up, even after core reset.

**Comment by birdybro** [comment]
It's not really a bug, it's that file extension detection does the following...

bin = NTSC-J
md = PAL
gen = NTSC

So if you don't have your NTSC ROM named "Game.gen" then you can't load it up in the right format.

The change in the latest unstable changes the default behavior for new users to header detection.

https://github.com/MiSTer-unstable-nightlies/Genesis_MiSTer/releases/download/unstable-builds/Genesis_unstable_20210922_7e1b.rbf

This latest unstable release fixes that and fixes an obscure bug with Star Trek TNG Echoes of the Past.

**Comment by divanikus** [comment]
I totally understand what your mean. But still, I think that could be considered a bug.

From my point of view, what happens? While loading a ROM core somehow sets up timings and other options for it according to a region it believes it belongs to. At very minimum they are resolution and refresh rate, might be something internal as well.

So when I change a region on-the-fly, I mean after ROM is already loaded, I do see it changes at least two things: resolution and screen refresh rate. Probably it doesn't change some internal stuff though. So core is running in some erroneous state, and this is why S&K is broken in that case.

So why it is even allowed to do so then? It shouldn't go into a not properly working, undocumented state. I think either region should be grayed out after loading, or the core should be able to properly switch to that region. Otherwise it is confusing for the end user.

**Comment by birdybro** [comment]
Fair points. If it were up to me I would remove File Ext altogether for the reasons you cited. I put Header as the default setting for the same reasons you are mentioning, that users get a bad experience with File Ext being the default.

```
It shouldn't go into a not properly working, undocumented state.
```

The File Ext setting is documented in the readme, it's just not intuitive and most people don't read readme's. Hopefully when a new release comes out and people reset their settings, less people will have an issue.

In order for the reset issue you have noticed to go away the rom would have to be reloaded upon pressing reset, which would take significantly more time. I think this would require a rewrite to Main_MiSTer and Template_MiSTer functionality.

**Comment by sorgelig** [comment]
Some multi-region games may detect region on startup only, so further change may not properly change all internal settings in game.

**Comment by divanikus** [comment]
If it is just a game code, why reset doesn't help though? Isn't it a kind of cold reboot?
```
The File Ext setting is documented in the readme
```
I meant the state after manual region change on the fly, not the `file ext` behavior itself.

**Comment by birdybro** [comment]
Reset is warm reboot.

Reloading the ROM is a cold reboot.

**Comment by Kitrinx** [comment]
the behavior would be similar to changing dip switches on the fly on something like a Scorpion console. It's not an impossible case, but if the region bothers you, simply reload the rom after changing it and you will get cold boot behavior.

**Comment by divanikus** [comment]
Cold reboot menu entry may be?

**Comment by birdybro** [comment]
Adding that consumes resources in the status bit map allocation unnecessarily if you can just reload rom already and achieve the same result.

**Comment by divanikus** [comment]
Well, I do understand your point. But just imaging how confusing it is for non advanced user.

**Comment by Kitrinx** [comment]
It's not very confusing. You load a rom, it's like you switched on the system. You choose reset, it resets the system. Changing region should probably send a cold boot signal and reset the game as well when it changes, I had thought that was already the case but maybe it was overlooked on that system.

**Comment by divanikus** [comment]
>  Changing region should probably send a cold boot signal and reset the game as well when it changes

Exactly my thoughts. I do not know if anyone really needs to change regions on the fly though.

**Comment by birdybro** [comment]
Right, what you experienced is due to header detection not being the default. This has been fixed in the latest commit, just a new release isn't out. The original hardware doesn't allow you to change regions on the fly and no game would require that to my knowledge. :)

**Comment by divanikus** [comment]
> Apparently there are a couple licensed games that take advantage of a warm reboot so this would cause a regression.

We aren't talking about replacing warm reboot with cold reboot, but sending a cold reboot on `Region` toggle in the user menu.

**Comment by birdybro** [comment]
Ah that's a good idea!

**Comment by Kitrinx** [comment]
there's no need for a toggle, this is implicit behavior, system should reset on system change

**Comment by divanikus** [comment]
> there's no need for a toggle, this is implicit behavior, system should reset on system change

I meant the action, toggling the region in the menu should emit a cold reboot signal. Idk if not doing so is a desired behavior for someone.

**Comment by birdybro** [comment]
Yeah, I'm looking into it personally now. Should be something like... removing the region set from line 473 of genesis.sv and adding a part for if region set is toggled then reload rom...

**Comment by birdybro** [comment]
Just randomly curious, @divanikus does this build resolve the issue?

[GenesisPALMCLKTest.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/7728657/GenesisPALMCLKTest.zip)

I doubt it will but wanted to see.

**Comment by divanikus** [comment]
> Just randomly curious, @divanikus does this build resolve the issue?

Sorry for the long delay with answer, I didn't had access to my MiSTer for a while. So, nope, It doesn't. If I start S3&K in EU region and later switch to US or JP, everything is glitching like hell.

**Comment by birdybro** [comment]
I tested this on Genesis-Plus-GX and the same bugs in Sonic 3 & Knuckles do not occur, so it's possible this should be ideally fixed elsewhere rather than a load rom being called upon on-the-fly region toggling. I also now suspect that it's related to the same Micro Machines EU etc... issues we have seen because the audio is distorted and it plays too fast. So many little obscure things that could just be bus arbiter related.
